prow/cmd/hmac/main.go

```diff
@@ -0,0 +1,389 @@
+	time.Sleep(20 * time.Second)
+	if err := c.batchOnboardNewTokenForRepos(); err != nil {
+		return fmt.Errorf("error onboarding new token for the repos: %v", err)
+	}
+
+	// Do necessary cleanups after the token and webhook updates are done.
+	if err := c.cleanup(); err != nil {
+		return fmt.Errorf("error cleaning up %v", err)
+	}
+
+	return nil
+}
+
+// handleRemoveRepo handles webhook removal and hmac token removal from the current hmac map for all repos removed from the declarative config.
+func (c *client) handleRemovedRepo(removed map[string]bool) error {
+	repos := make([]string, len(removed))
+	var i int
+	for k := range removed {
+		repos[i] = k
+		i++
+	}
+
+	o := ghhook.Options{
+		GitHubOptions: c.options.github,
+		Repos:         prowflagutil.NewStrings(repos...),
+		HookURL:       c.options.hookUrl,
+		ShouldDelete:  true,
+		Confirm:       true,
+	}
+	if err := o.Initialize(); err != nil {
+		return err
+	}
+
+	if err := o.HandleWebhookConfigChange(); err != nil {
+		return fmt.Errorf("error deleting webhook for repos %q: %v", repos, err)
+	}
+	for _, repo := range repos {
+		delete(c.currentHMACMap, repo)
+	}
+	// No need to update the secret here, the following update will commit the changes together.
+
+	return nil
+}
+
+func (c *client) handleAddedRepo(added map[string]config.ManagedWebhookInfo) error {
+	for repo := range added {
+		if err := c.addRepoToBatchUpdate(repo); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *client) handledRotatedRepo(rotated map[string]config.ManagedWebhookInfo) error {
+	// For each rotated repo, we only onboard a new token when none of the existing tokens is created after user specified time.
+	for repo, hmacConfig := range rotated {
+		needsRotation := true
+		for _, token := range c.currentHMACMap[repo] {
+			// If the existing token is created after the user specified time, we do not need to rotate it.
+			if token.CreatedAt.After(hmacConfig.TokenCreatedAfter) {
+				needsRotation = false
+				break
+			}
+		}
+		if needsRotation {
+			if err := c.addRepoToBatchUpdate(repo); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *client) addRepoToBatchUpdate(repo string) error {
+	generatedToken, err := generateNewHmacToken()
+	if err != nil {
+		return fmt.Errorf("error generating a new hmac token for %q: %v", repo, err)
+	}
+
+	updatedTokenList := github.HmacsForRepo{}
+	if val, ok := c.currentHMACMap[repo]; ok {
+		// Copy over all existing tokens for that repo.
+		updatedTokenList = append(updatedTokenList, val...)
+	} else {
+		// Current webhook is possibly using global token so we need to promote that token to repo level.
+		globalTokens := c.currentHMACMap["*"]
+		updatedTokenList = append(updatedTokenList, github.HmacSecret{
+			Value: globalTokens[0].Value,
+			// Set CreatedAt as a time slightly before now, so that the token can be properly pruned in the end.
+			CreatedAt: time.Now().Add(-time.Hour),
```

Comment:
This should be before the `TokenCreatedAfter` time, not before now. An hour before now would probably work in practice since `TokenCreatedAfter` should be approximately the current time, but 1 second (or millisecond) before the `TokenCreatedAfter` time is what we really need to ensure proper pruning.

Reply:
Done, changed to `c.newHMACConfig[repo].TokenCreatedAfter.Add(-time.Second)`.